flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java

```diff
@@ -124,17 +154,33 @@ private void handleDeploymentFailure(ExecutionVertex executionVertex, JobExcepti
 
     @Override
     public void onNewResourcesAvailable() {
-        maybeRescale();
+        rescaleWhenCooldownPeriodIsOver();
     }
 
     @Override
     public void onNewResourceRequirements() {
-        maybeRescale();
+        rescaleWhenCooldownPeriodIsOver();
     }
 
     private void maybeRescale() {
-        if (context.shouldRescale(getExecutionGraph())) {
-            getLogger().info("Can change the parallelism of job. Restarting job.");
+        final Duration timeSinceLastRescale = timeSinceLastRescale();
+        rescaleScheduled = false;
+        final boolean shouldForceRescale =
+                (scalingIntervalMax != null)
+                        && (timeSinceLastRescale.compareTo(scalingIntervalMax) > 0)
+                        && (lastRescale != Instant.EPOCH); // initial rescale is not forced
+        if (shouldForceRescale || context.shouldRescale(getExecutionGraph())) {
+            if (shouldForceRescale) {
+                getLogger()
+                        .info(
+                                "Time since last rescale ({}) >  {} ({}). Force-changing the parallelism of the job. Restarting the job.",
+                                timeSinceLastRescale,
+                                JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MAX.key(),
+                                scalingIntervalMax);
+            } else {
+                getLogger().info("Can change the parallelism of the job. Restarting the job.");
+            }
+            lastRescale = Instant.now();
             context.goToRestarting(
                     getExecutionGraph(),
```

Comment:
The implementation of  `scalingIntervalMax` may be wrong.

As I understand, `scalingIntervalMax` want to trigger a rescale when the `jobmanager.adaptive-scheduler.min-parallelism-increase` isn't met, right?

If yes, we should set a timer to call the `rescale()` after `scalingIntervalMax`, right? If we don't set a timer, the `rescale` won't be called after `scalingIntervalMax`, and no one to force rescale.

Also, we should add a test for this case.

Reply:
in the FLIP we said that the triggering event of the rescale is resource addition (`Executing#onNewResourcesAvailable`) (downscale is triggered by failures) so there is no need to schedule: when such an event happens `maybeRescale` is always called (either directly or scheduled) and when it is called, the rescale is done if `timeSinceLastRescale > scalingIntervalMax` no matter the min resources.  

There is also a test: `ExecutingTest#testNotifyNewResourcesAvailableWithNoResourcesAndForce`

Reply:
Thanks for the clarification!

Do you mean we force trigger a rescale when new resources comes and the `current time > lastRescaleTime + scalingIntervalMax`?

If so, I think the logic related to `scalingIntervalMax` should be moved into `maybeRescale`. `onNewResourceRequirements` also call the `rescaleWhenCooldownPeriodIsOver`, when user changed the parallelism, `maybeRescale` will force trigger a rescale as well. right?

Reply:
> Do you mean we force trigger a rescale when new resources comes and the current time > lastRescaleTime + scalingIntervalMax?

yes, that is what is what the community agreed on in the FLIP.

> If so, I think the logic related to scalingIntervalMax should be moved into maybeRescale.

This logic is already in maybeRescale() method

> onNewResourceRequirements also call the rescaleWhenCooldownPeriodIsOver, when user changed the parallelism; maybeRescale will force trigger a rescale as well. right?

yes onNewResourceRequirements is ultimately called by a rest endpoint to react to use action on the web ui so it will call maybeRescale in the end but that would not always mean a rescale: if timeSinceLastRescale < scalingIntervalMax and minimal resources are not met then there will be no rescale.

Reply:
> > If so, I think the logic related to scalingIntervalMax should be moved into maybeRescale.
> 
> This logic is already in maybeRescale() method

Sorry, I have a typo here. I mean the logic related to scalingIntervalMax should be moved out of the maybeRescale. 

`onNewResourceRequirements` also call the `rescaleWhenCooldownPeriodIsOver` when user changed the parallelism, maybeRescale shouldn't force trigger a rescale in this case even if the `current time > lastRescaleTime + scalingIntervalMax`.

Because after requirement is adjusted, resources cannot be ready immediately. The rescale is meaningless at this time even if the `current time > lastRescaleTime + scalingIntervalMax`.

Reply:
> > > If so, I think the logic related to scalingIntervalMax should be moved into maybeRescale.
> > 
> > 
> > This logic is already in maybeRescale() method
> 
> Sorry, I have a typo here. I mean the logic related to scalingIntervalMax should be moved out of the maybeRescale.
> 
> `onNewResourceRequirements` also call the `rescaleWhenCooldownPeriodIsOver` when user changed the parallelism, maybeRescale shouldn't force trigger a rescale in this case even if the `current time > lastRescaleTime + scalingIntervalMax`.
> 
> Because after requirement is adjusted, resources cannot be ready immediately. The rescale is meaningless at this time even if the `current time > lastRescaleTime + scalingIntervalMax`.

Agree, I think the same is true for maybeRescale() called in the constructor to check if new resources have arrived during the starting process

Reply:
> > > Do you mean we force trigger a rescale when new resources comes and the current time > lastRescaleTime + scalingIntervalMax?
> > 
> > 
> > yes, that is what is what the community agreed on in the FLIP.
> 
> I think this strategy just solve one case: the last resource comes after scalingIntervalMax. The job can rescale directly.
> 

It was made after this observation, the longer the pipeline runs, the more the resource gain is worth the restarting time. 

> However, it cannot solve the case: the last resource comes before scalingIntervalMax. Assuming the scalingIntervalMax is 5 miuntes, the `jobmanager.adaptive-scheduler.min-parallelism-increase` is 2, and the expected paralleslim is 100.
> 
> * The job has 99 TMs at 09:00:00, it run with parallelism 99.
> * At 09:03:00, the last TM comes.
> * The `Executing` will ignore the rescale due to the parallelism diff is 1, it less than `min-parallelism-increase` and `current time < lastRescaleTime + scalingIntervalMax`.
> 
> And then the job doesn't need the new resource and the expected parallelism isn't change, so the job won't rescale anymore.

It is not linked to the scalingIntervalMax, before this PR in such a case there was no rescale either because simply the minimum resource addition was not met.  It is by design of the min-parallelism-increase that we don't change

Reply:
> It is by design of the min-parallelism-increase that we don't change

This design was changed for case1, right? 

If yes, the `scalingIntervalMax` wants to ignore the `min-parallelism-increase` if one job doesn't rescale for a long time. I think the semantic of scalingIntervalMax should be fixed to the user side and should not be related to the calling order in the code or the order in which resources arrive.

Reply:
> I think the semantic of scalingIntervalMax should be fixed to the user side and should not be related to the calling order in the code or the order in which resources arrive.

I don't understand what you mean by "fixed to the user side".

Reply:
Here is the summary of the offline discussion @1996fanrui and I had about `scalingIntervalMax` and forcing rescale (current code: when `added resources < min-parallelism-increase` we force a rescale if `timeSinceLastRescale > scalingIntervalMax`):
1. aim: the longer the pipeline runs, the more the (small) resource gain is worth the restarting time.
2. corner case: a resource `< min-parallelism-increase` arrives when `timeSinceLastRescale < scalingIntervalMax` and the pipeline is running for a long time (typical case 1) => with the current code, we don't force a rescale in that case whereas the added resource would be worth the restarting time.
=> I proposed solution 1: changing the definition of  `scalingIntervalMax` to `pipelineRuntimeRescaleMin` meaning pipeline runtime after which we force a rescale even if `added resources < min-parallelism-increase`
=> @1996fanrui  proposed solution 2: if `added resources < min-parallelism-increase && timeSinceLastRescale < scalingIntervalMax` schedule a tryRescale (force a rescale if there is indeed a change in the resource graph at that time in case the last TM crashed) after `scalingIntervalMax`

@zentol you were the one who proposed the addition of `scalingIntervalMax` in the FLIP discussion thread. Do you prefer solution 1 or solution 2 ?

Reply:
I don't really see a difference between the 2 options. In both options you had a running job, got resources that do not exceed min-parallellism-increase, and trigger a rescale later after scalingIntervalMax/pipelineRuntimeRescaleMin (or however it is called).

It just seems like option 2 applies scalingIntervalMax twice; once as a condition to schedule tryRescaled, and another time by using it as the delay.

But I assume that that's a mistake in the description, and the difference is meant to be _when_ you start the timeout.
Option 1 does it when a resource is acquired, the 2nd option after a rescale.

So, for a concrete example: We have a job running at p=1 for 24h straight, have a min increase of 1, and a timeout of 1h.
Now we get another slot at the 24h mark.

With option 1 we rescale after an hour.
With option 2 we rescale right away because the timeout already elapsed within the previous 24 hours.

Option 2 is pretty clever, but I think it might be side-stepping the intentions behind the min increase option and timeout.

The min increase option was added to prevent frequent rescalings when resources trickle in slowly.
A slot arrives, you rescale. Another one arrives 5 minutes later, you rescale again, and so on.
The timeout was meant for the case where no further resources arrive; you're holding on to this 1 slot _forever_, but using it would pay off over time.

If you rescale immediately, you will end up rescaling twice in the case where multiples slots do arrive slowly.

With that in mind I'd still prefer option 1.

Reply:
Hi @zentol , thanks for your feedback here.

Maybe you misunderstand here. I just clarify what the difference between @echauchot  and me. The current uncertainty is which cases we want to solve rather than the option or how the code is implemented. Here is a case: 

- Assuming the scalingIntervalMax is 60 miuntes, the `jobmanager.adaptive-scheduler.min-parallelism-increase` is 2, and the expected paralleslim is 100.
- The job starts at 09:00:00, and it run with parallelism 99.
- At 09:40:00, the last TM comes.

For this case, the runningTime is 40 minutes, it's less than scalingIntervalMax(60 miuntes). And the resource diff is 1 , it's less than 2 (`jobmanager.adaptive-scheduler.min-parallelism-increase` ).

- @echauchot prefer ignore this case, and never rescale.
- I prefer cover this case, the solution is schedule a rescale after `scalingIntervalMax` or `scalingIntervalMax - runningTime`.

The solution can be ignored now, what do you think about this case? Should we cover this case?

@echauchot Please correct me if I'm wrong, thanks~

Reply:
Thanks Chesnay for your views (again) ! 
> With option 2 we rescale right away because the timeout already elapsed within the previous 24 hours.

This is what I proposed indeed, but you're right that

> it might be side-stepping the intentions behind the min increase option and timeout

if for example after the 24h resources start to arrive 1 slot every 5 min, with this scenario we will restart every 5 min (as the timeout is exceeded) which we want to avoid. So it is better to schedule a timeout when resources arrive and `added ressource < min increase`. When the timeout fires we do one single rescale that takes all the added slots in one shot.
I'll do that

Reply:
> Should we cover this case?

We should absolutely cover that case. The whole `scalingIntervalMax` idea was meant for that exact scenario; you have free slots below the min-parallelism-increase threshold for a long time. If you don't cover that case you just have the cooldown between rescale operations.

Reply:
>  with this scenario we will restart every 5 min (as the timeout is exceeded min-increase check is overridden) 

This depends on what your starting point for the timeout is. It shouldn't be fixed to the start of the job, but be reset after a job was rescaled.
So, slot arrives at 24h mark, job is rescaled, timeout is reset; for the next hour we only scale up if min-parallel-increase is satisfied, if a slot arrives after that we'd immediately rescale.
I still think this'd be problematic though as described above.


I think a good way to think about it is like this: We probably all had some situation in our live where we thought "oh If only I had done that <some time> earlier." Can be anything, really, let's say it's washing the dishes. You can still do that _now_, but had you done it earlier you may _now_ have been able to use your time better.
That's ultimately what we're doing here.
We look back, see that we had this slot the whole time, and go "ah, if only I had rescaled sooner, but better late than never".
But on the flip-side, if you lost that slot in the mean time and look back you'd think "good job me, that was a good call to ignore that slot!".

This is all about stability; "I could have rescaled and it wouldn't have been a problem.". But you can only conclude that after having out that slot for some period of time; otherwise you're just guessing.

Reply:
> This is all about stability; "I could have rescaled and it wouldn't have been a problem.". But you can only conclude that after having out that slot for some period of time; otherwise you're just guessing.

Exactly. So, when slot arrives scheduling a rescale-check after a timeout seems a good trade-off to ensure that the slot is still there after a period of time. When the timeout fires:
- If the slot is still here we should force-rescale (better late than never case)
- If the slot has vanished we should determine that the target parallelism is the same as current and not rescale (avoid unnecessary rescale case)

Reply:
done